File: androidApp/core/compose/src/main/kotlin/com/velkonost/getbetter/core/compose/components/note/subnotes/SubNotesBlock.kt
package com.velkonost.getbetter.core.compose.components.note.subnotes

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import com.velkonost.getbetter.core.compose.components.PrimaryBox
import com.velkonost.getbetter.core.compose.theme.Dimen.DP_16
import com.velkonost.getbetter.core.compose.theme.Dimen.DP_6
import com.velkonost.getbetter.core.compose.theme.Pixel.PX_ZERO
import com.velkonost.getbetter.shared.core.model.ui.SubNoteUI
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.stringResource

@Composable
fun SubNotesBlock(
    modifier: Modifier = Modifier,
    onlyView: Boolean = false,
    newSubNote: SubNoteUI,
    items: List<SubNoteUI>,
    isCompleteVisible: Boolean = false,
    isSubNotesBlockVisible: MutableState<Boolean>,
    onNewSubNoteChanged: (String) -> Unit,
    onAddNewSubNote: () -> Unit,
    onSubNoteDelete: (SubNoteUI) -> Unit,
    onCompleteClick: ((SubNoteUI) -> Unit)? = null
) {

    val itemPadding = remember { DP_6 }
    val viewPadding = remember { PX_ZERO }
    val contentBottomPadding = remember { DP_16 }

    PrimaryBox(padding = viewPadding) {
        Column {
            SubNotesHeader(isSubNotesBlockVisible = isSubNotesBlockVisible)

            AnimatedVisibility(visible = isSubNotesBlockVisible.value) {
                AnimatedContent(targetState = items, label = "") { subNotes ->
                    Column(
                        modifier = modifier.padding(bottom = contentBottomPadding),
                        verticalArrangement = Arrangement.spacedBy(itemPadding),
                    ) {
                        subNotes.forEach { item ->
                            SubNoteItem(
                                item = item,
                                onlyView = onlyView,
                                isCompleteVisible = isCompleteVisible,
                                onDeleteSubNote = onSubNoteDelete,
                                onCompleteClick = onCompleteClick
                            )
                        }

                        AnimatedVisibility(visible = !onlyView) {
                            AddSubNoteItem(
                                value = newSubNote.text,
                                placeholderText = stringResource(resource = SharedR.strings.create_note_subnote_hint),
                                onValueChanged = onNewSubNoteChanged,
                                onAddSubNote = onAddNewSubNote
                            )
                        }

                    }
                }

            }
        }
    }

}

File: iosApp/iosApp/Core/Components/note/subnotes/SubNotesBlock.swift
//
//  SubNotesBlock.swift
//  iosApp
//
//  Created by velkonost on 26.10.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK

struct SubNotesBlock: View {
    
    let items: [SubNoteUI]
    @Binding var newSubNote: SubNoteUI
    
    let onlyView: Bool
    let isCompleteVisible: Bool
    
    let onNewSubNoteChanged: (String) -> Void
    let onAddNewSubNote: () -> Void
    let onSubNoteDelete: (SubNoteUI) -> Void
    let onCompleteClick: ((SubNoteUI) -> Void)?
    
    @Binding private var isSubNotesBlockPickerVisible: Bool
    
    init(items: [SubNoteUI], newSubNote: Binding<SubNoteUI>, onlyView: Bool = false, isCompleteVisible: Bool = false, onNewSubNoteChanged: @escaping (String) -> Void, onAddNewSubNote: @escaping () -> Void, onSubNoteDelete: @escaping (SubNoteUI) -> Void, isSubNotesBlockPickerVisible: Binding<Bool>, onCompleteClick: ((SubNoteUI) -> Void)? = nil) {
        self.items = items
        self._newSubNote = newSubNote
        self.onlyView = onlyView
        self.isCompleteVisible = isCompleteVisible
        self.onNewSubNoteChanged = onNewSubNoteChanged
        self.onAddNewSubNote = onAddNewSubNote
        self.onSubNoteDelete = onSubNoteDelete
        self.onCompleteClick = onCompleteClick
        self._isSubNotesBlockPickerVisible = isSubNotesBlockPickerVisible
    }
    
    var body: some View {
        PrimaryBox(
            padding: .init(top: .zero, leading: .zero, bottom: .zero, trailing: .zero)
        ) {
            VStack {
                SubNotesHeader(
                    isSubNotesBlockVisible: isSubNotesBlockPickerVisible,
                    onClick: {
                        withAnimation {
                            isSubNotesBlockPickerVisible.toggle()
                        }
                    }
                )
                
                if isSubNotesBlockPickerVisible {
                    LazyVStack {
                        ForEach(items, id: \.self.id) { item in
                            SubNoteItem(
                                item: item,
                                onlyView: onlyView,
                                isCompleteVisible: isCompleteVisible,
                                onDeleteSubNote: onSubNoteDelete,
                                onCompleteClick: onCompleteClick
                            )
                        }
                        
                        if !onlyView {
                            AddSubNoteItem(
                                value: $newSubNote,
                                placeholderText: SharedR.strings().create_note_subnote_hint.desc().localized(),
                                onValueChanged: onNewSubNoteChanged,
                                onAddSubNote: onAddNewSubNote
                            )
                        }
                    }
                    .padding(.leading, 16)
                    .padding(.trailing, 16)
                    
                }
            }
            .animation(.easeInOut, value: onlyView)
            .animation(.easeInOut, value: isCompleteVisible)
            .padding(.bottom, 16)
            
        }
        .animation(.easeInOut, value: items)
    }
}
